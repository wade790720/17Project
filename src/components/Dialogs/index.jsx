import React from 'react';
import PropTypes from 'prop-types';
import "./style/dialogs.sass"
import Button from "components/Button";
import Item from "components/Item";
import { uniqueKey } from "utils/uniqueKey";

const defaultItems = {
    id: uniqueKey(),
    name: '',
    price: 0,
    remark: ''
}

class Dialogs extends React.Component {
    state = { ...this.props.order }

    componentDidMount() {
        const order = localStorage.getItem(this.props.orderId)
        if (order) this.setState({ items: JSON.parse(order) });
    }

    componentWillUnmount() {
        this.clear();
    }

    addItem = () => {
        this.setState({
            items: [
                ...this.state.items,
                {
                    id: uniqueKey(),
                    name: '',
                    price: 0,
                    remark: ''
                }
            ]
        });
    }

    removeItem = (id) => {
        this.setState({
            items: this.state.items.filter(item => item.id !== id)
        });
    }

    save = () => {
        this.props.onSave(this.state);
    }

    clear = () => {
        this.setState({
            id: 0,
            storeName: "",
            items: [{ ...defaultItems }]
        });
    }

    handleChangeStoreName = (e) => {
        this.setState({ storeName: e.target.value });
    }

    handleChangeItem = (newItem) => {
        this.setState({
            items: this.state.items.map(item => (item.id === newItem.id ? newItem : item))
        });
    }

    render() {
        const { storeName, items } = this.state;
        return (
            <div className="dialogs open">
                <div className="dialogs-box">
                    <div className="title">
                        <input type="text" placeholder="Store Name" value={storeName} onChange={this.handleChangeStoreName} />
                    </div>
                    <div className="content">
                        {items.map(item =>
                            <Item
                                key={item.id}
                                item={item}
                                onChange={this.handleChangeItem}
                                onDelete={() => this.removeItem(item.id)}
                            />
                        )}
                    </div>
                    <div className="action">
                        <Button className="success" text="新增" onClick={this.addItem} />
                        <Button className="primary" text="儲存" onClick={this.save} />
                        <Button className="default" text="關閉" onClick={this.props.onCancel} />
                    </div>
                </div>
            </div>
        )
    }
};

export default Dialogs;

Dialogs.defaultProps = {
    order: {
        id: '',
        storeName: '',
        items: [defaultItems],
    },
    onSave: () => { },
    onCancel: () => { }
}
Dialogs.propTypes = {
    order: PropTypes.object,
    onSave: PropTypes.func,
    onCancel: PropTypes.func
}