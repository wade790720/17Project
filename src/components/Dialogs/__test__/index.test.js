import React from "react";
import { shallow } from "enzyme";
import Dialogs from "../index";
import renderer from "react-test-renderer";

const fakeOrder = {
    id: '',
    storeName: '',
    items: [{ id: '1', name: '', price: 0, remark: '' }]
}

describe('Dialogs tests', () => {
    it("Default view", () => {
        const tree = renderer.create(
            <Dialogs />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("no data view", () => {
        const order = { id: 'abc', storeName: 'abb', items: [] }
        const tree = renderer.create(
            <Dialogs order={order} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('addItem function test', () => {
        const component = shallow(<Dialogs />);
        component.instance().addItem();
        expect(component.instance().state.items.length).toBe(2);
        expect(component.instance().state.items[0].id).not.toBe("");
        expect(component.instance().state.items[1].id).not.toBe("");
        expect(component.instance().state.items[0]).toEqual(expect.objectContaining({
            id: expect.any(String),
            name: expect.any(String),
            price: expect.any(Number),
            remark: expect.any(String)
        }));
    });

    it('removeItem function test', () => {
        const component = shallow(<Dialogs order={fakeOrder} />);
        expect(component.instance().state.items.length).toBe(1)
        component.instance().removeItem("1");
        expect(component.instance().state.items.length).toBe(0)
    });

    it('handleChangeStoreName function test', () => {
        const component = shallow(<Dialogs />);
        component.find(".title input").simulate("change", {
            target: { value: 'StoreName' }
        });
        expect(component.instance().state.storeName).toBe("StoreName");
    });

    it('handleChangeItem function test', () => {
        const component = shallow(<Dialogs order={fakeOrder} />);
        const data = {
            id: '1',
            name: 'wade',
            price: 0,
            remark: ''
        };
        component.instance().handleChangeItem(data);
        expect(component.instance().state.items[0].name).toBe(data.name);
    });
});