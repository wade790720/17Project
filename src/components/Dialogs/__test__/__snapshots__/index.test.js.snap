// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogs tests Default view 1`] = `
<div
  className="dialogs open"
>
  <div
    className="dialogs-box"
  >
    <div
      className="title"
    >
      <input
        onChange={[Function]}
        placeholder="Store Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="content"
    >
      <div
        className="item"
      >
        <input
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Name"
          type="text"
          value=""
        />
        <input
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Price"
          type="number"
          value={0}
        />
        <input
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Remark"
          type="text"
          value=""
        />
        <div
          className="remove"
          onClick={[Function]}
        />
      </div>
    </div>
    <div
      className="action"
    >
      <button
        className="btn success"
        onClick={[Function]}
      >
        新增
      </button>
      <button
        className="btn primary"
        onClick={[Function]}
      >
        儲存
      </button>
      <button
        className="btn default"
        onClick={[Function]}
      >
        關閉
      </button>
    </div>
  </div>
</div>
`;

exports[`Dialogs tests no data view 1`] = `
<div
  className="dialogs open"
>
  <div
    className="dialogs-box"
  >
    <div
      className="title"
    >
      <input
        onChange={[Function]}
        placeholder="Store Name"
        type="text"
        value="abb"
      />
    </div>
    <div
      className="content"
    />
    <div
      className="action"
    >
      <button
        className="btn success"
        onClick={[Function]}
      >
        新增
      </button>
      <button
        className="btn primary"
        onClick={[Function]}
      >
        儲存
      </button>
      <button
        className="btn default"
        onClick={[Function]}
      >
        關閉
      </button>
    </div>
  </div>
</div>
`;
